sys/kern/subr_intr.c

```diff
@@ -588,131 +703,6 @@ iscr_setup_filter(struct intr_irqsrc *isrc, const char *name,
 }
 #endif
 
-/*
- *  Interrupt source pre_ithread method for MI interrupt framework.
- */
-static void
-intr_isrc_pre_ithread(void *arg)
-{
-	struct intr_irqsrc *isrc = arg;
-
-	PIC_PRE_ITHREAD(isrc->isrc_dev, isrc);
-}
-
-/*
- *  Interrupt source post_ithread method for MI interrupt framework.
- */
-static void
-intr_isrc_post_ithread(void *arg)
-{
-	struct intr_irqsrc *isrc = arg;
-
-	PIC_POST_ITHREAD(isrc->isrc_dev, isrc);
-}
-
-/*
- *  Interrupt source post_filter method for MI interrupt framework.
- */
-static void
-intr_isrc_post_filter(void *arg)
-{
-	struct intr_irqsrc *isrc = arg;
-
-	PIC_POST_FILTER(isrc->isrc_dev, isrc);
-}
-
-/*
- *  Interrupt source assign_cpu method for MI interrupt framework.
- */
-static int
-intr_isrc_assign_cpu(void *arg, int cpu)
-{
-#ifdef SMP
-	struct intr_irqsrc *isrc = arg;
-	int error;
-
-	mtx_lock(&isrc_table_lock);
-	if (cpu == NOCPU) {
-		CPU_ZERO(&isrc->isrc_cpu);
-		isrc->isrc_flags &= ~INTR_ISRCF_BOUND;
-	} else {
-		CPU_SETOF(cpu, &isrc->isrc_cpu);
-		isrc->isrc_flags |= INTR_ISRCF_BOUND;
-	}
-
-	/*
-	 * In NOCPU case, it's up to PIC to either leave ISRC on same CPU or
-	 * re-balance it to another CPU or enable it on more CPUs. However,
-	 * PIC is expected to change isrc_cpu appropriately to keep us well
-	 * informed if the call is successful.
-	 */
-	if (irq_assign_cpu) {
-		error = PIC_BIND_INTR(isrc->isrc_dev, isrc);
-		if (error) {
-			CPU_ZERO(&isrc->isrc_cpu);
-			mtx_unlock(&isrc_table_lock);
-			return (error);
-		}
-	}
-	mtx_unlock(&isrc_table_lock);
-	return (0);
-#else
-	return (EOPNOTSUPP);
-#endif
-}
-
-/*
- *  Create interrupt event for interrupt source.
- */
-static int
-isrc_event_create(struct intr_irqsrc *isrc)
-{
-	struct intr_event *ie;
-	int error;
-
-	error = intr_event_create(&ie, isrc, 0, isrc->isrc_irq,
-	    intr_isrc_pre_ithread, intr_isrc_post_ithread, intr_isrc_post_filter,
-	    intr_isrc_assign_cpu, "%s:", isrc->isrc_name);
-	if (error)
-		return (error);
-
-	mtx_lock(&isrc_table_lock);
-	/*
-	 * Make sure that we do not mix the two ways
-	 * how we handle interrupt sources. Let contested event wins.
-	 */
-#ifdef INTR_SOLO
-	if (isrc->isrc_filter != NULL || isrc->isrc_event != NULL) {
-#else
-	if (isrc->isrc_event != NULL) {
-#endif
-		mtx_unlock(&isrc_table_lock);
-		intr_event_destroy(ie);
-		return (isrc->isrc_event != NULL ? EBUSY : 0);
-	}
-	isrc->isrc_event = ie;
-	mtx_unlock(&isrc_table_lock);
-
-	return (0);
-}
-#ifdef notyet
-/*
- *  Destroy interrupt event for interrupt source.
- */
-static void
-isrc_event_destroy(struct intr_irqsrc *isrc)
-{
-	struct intr_event *ie;
-
-	mtx_lock(&isrc_table_lock);
-	ie = isrc->isrc_event;
-	isrc->isrc_event = NULL;
-	mtx_unlock(&isrc_table_lock);
-
-	if (ie != NULL)
-		intr_event_destroy(ie);
-}
-#endif
```

Comment:
Why is all this code moving around?

Reply:
The issue is all of these are `static` without separate declarations.  As such `intr_isrc_register()` calling `isrc_event_create()` means `isrc_event_create()` needs to be above `intr_isrc_register()`.  `isrc_event_destroy()` is moved to keep it next to `isrc_event_create()`.

In turn `isrc_event_create()` passes pointers to `intr_isrc_pre_ithread()`, `intr_isrc_post_ithread()`, `intr_isrc_post_filter()` and `intr_isrc_assign_cpu()` to `intr_event_create()`.  Since those four have no declarations, they must move earlier in the file than `isrc_event_create()`.